File: packages/scene-schema/src/scene.ts
import { z } from 'zod';

import { audioBindingSchema } from './audio';

export const SCENE_SCHEMA_VERSION = '1.0' as const;

export const presetNameSchema = z.enum([
  'soft-nature',
  'urban',
  'dark',
  'electronic',
  'nostalgic',
]);
export type PresetName = z.infer<typeof presetNameSchema>;

export const layerRoleSchema = z.enum([
  'background',
  'distant',
  'midground',
  'secondary-subject',
  'primary-subject',
  'foreground',
  'atmosphere',
]);
export type LayerRole = z.infer<typeof layerRoleSchema>;

export const returnModeSchema = z.enum(['spring', 'inertia', 'physics', 'fixed']);
export type ReturnMode = z.infer<typeof returnModeSchema>;

/** Normalized [0,1] rectangle in image space (origin top-left). */
export const boundsSchema = z.object({
  x: z.number(),
  y: z.number(),
  width: z.number().positive(),
  height: z.number().positive(),
});
export type Bounds = z.infer<typeof boundsSchema>;

export const anchorSchema = z.object({
  x: z.number(),
  y: z.number(),
});

export const layerMovementSchema = z.object({
  enabled: z.boolean(),
  maxOffsetX: z.number().min(0),
  maxOffsetY: z.number().min(0),
  maxOffsetZ: z.number().min(0),
  maxRotation: z.number().min(0),
  parallaxStrength: z.number().min(0),
  dragEnabled: z.boolean(),
  returnMode: returnModeSchema,
});
export type LayerMovement = z.infer<typeof layerMovementSchema>;

export const audioSensitivitySchema = z.object({
  bass: z.number().min(0).max(1),
  lowMid: z.number().min(0).max(1),
  highMid: z.number().min(0).max(1),
  treble: z.number().min(0).max(1),
  beat: z.number().min(0).max(1),
  loudness: z.number().min(0).max(1),
});
export type AudioSensitivity = z.infer<typeof audioSensitivitySchema>;

/** Limits on how far a layer may move, so we never reveal unreconstructed pixels. */
export const revealBudgetSchema = z.object({
  maxOffsetX: z.number().min(0),
  maxOffsetY: z.number().min(0),
  confidence: z.number().min(0).max(1),
  generatedCoverageMaskUrl: z.string().optional(),
});
export type RevealBudget = z.infer<typeof revealBudgetSchema>;

export const provenanceSchema = z.object({
  visiblePixels: z.enum(['original', 'generated', 'mixed']),
  sourceImageHash: z.string(),
  segmentationProvider: z.string().optional(),
  mattingProvider: z.string().optional(),
  completionProvider: z.string().optional(),
  providerRequestIds: z.array(z.string()).optional(),
});
export type Provenance = z.infer<typeof provenanceSchema>;

export const sceneLayerSchema = z.object({
  id: z.string().min(1),
  name: z.string(),
  semanticLabel: z.string(),
  role: layerRoleSchema,

  assetUrl: z.string().min(1),
  maskUrl: z.string().optional(),
  alphaAssetUrl: z.string().optional(),

  bounds: boundsSchema,
  anchor: anchorSchema,

  /** Relative depth in [0,1]; 0 = nearest camera, 1 = farthest. */
  depth: z.number().min(0).max(1),
  depthVariance: z.number().min(0),
  baseScale: z.number().positive(),
  baseRotation: z.number(),
  baseOpacity: z.number().min(0).max(1),

  movement: layerMovementSchema,

  interactionTags: z.array(z.string()),
  materialTags: z.array(z.string()),
  audioSensitivity: audioSensitivitySchema,

  importance: z.number().min(0).max(1),
  locked: z.boolean(),

  revealBudget: revealBudgetSchema,
  provenance: provenanceSchema,
});
export type SceneLayer = z.infer<typeof sceneLayerSchema>;

export const visualAnalysisSchema = z.object({
  sceneType: z.string(),
  mainSubject: z.string(),
  secondarySubjects: z.array(z.string()).default([]),
  foregroundCandidates: z.array(z.string()).default([]),
  midgroundCandidates: z.array(z.string()).default([]),
  background: z.string().default(''),
  waterRegions: z.array(z.string()).default([]),
  vegetationRegions: z.array(z.string()).default([]),
  skyRegions: z.array(z.string()).default([]),
  transparentOrReflectiveRegions: z.array(z.string()).default([]),
  lightingDirection: z.string().default(''),
  depthOfField: z.string().default(''),
  cameraStyle: z.string().default(''),
  dominantColors: z.array(z.string()).default([]),
  mood: z.string().default(''),
  suggestedPreset: presetNameSchema.optional(),
  suggestedMotionTags: z.array(z.string()).default([]),
  suggestedSegmentationPrompts: z.array(z.string()).default([]),
  lockedObjects: z.array(z.string()).default([]),
  expectedSegmentationDifficulty: z.enum(['low', 'medium', 'high']).default('medium'),
  inpaintingRiskAreas: z.array(z.string()).default([]),
});
export type VisualAnalysis = z.infer<typeof visualAnalysisSchema>;

export const atmosphereConfigSchema = z.object({
  fog: z
    .object({
      enabled: z.boolean().default(false),
      density: z.number().min(0).max(1).default(0),
      color: z.string().default('#ffffff'),
      layers: z.number().int().min(0).max(5).default(0),
    })
    .default({}),
  ambientLight: z.number().min(0).max(2).default(1),
  vignette: z.number().min(0).max(1).default(0),
  grain: z.number().min(0).max(1).default(0),
});
export type AtmosphereConfig = z.infer<typeof atmosphereConfigSchema>;

export const cameraConfigSchema = z.object({
  /** Field of view in degrees for the perspective camera. */
  fov: z.number().min(1).max(179).default(45),
  parallaxStrength: z.number().min(0).default(0.16),
  zoomResponse: z.number().min(0).default(0.03),
  shakeStrength: z.number().min(0).default(0),
  driftStrength: z.number().min(0).default(0),
});
export type CameraConfig = z.infer<typeof cameraConfigSchema>;

export const sceneMetadataSchema = z.object({
  createdAt: z.string(),
  pipelineVersion: z.string(),
});

export const sceneDocumentSchema = z
  .object({
    version: z.literal(SCENE_SCHEMA_VERSION),
    id: z.string().min(1),
    title: z.string(),
    width: z.number().int().positive(),
    height: z.number().int().positive(),
    aspectRatio: z.number().positive(),
    originalImageUrl: z.string().min(1),
    backgroundPlateUrl: z.string().min(1),
    depthMapUrl: z.string().optional(),
    preset: presetNameSchema,
    visualAnalysis: visualAnalysisSchema,
    layers: z.array(sceneLayerSchema).min(1),
    atmosphere: atmosphereConfigSchema,
    camera: cameraConfigSchema,
    audioBindings: z.array(audioBindingSchema).default([]),
    metadata: sceneMetadataSchema,
  })
  .superRefine((doc, ctx) => {
    // Aspect ratio must be internally consistent (within rounding tolerance).
    const expected = doc.width / doc.height;
    if (Math.abs(expected - doc.aspectRatio) > 0.02) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['aspectRatio'],
        message: `aspectRatio ${doc.aspectRatio} does not match width/height (${expected.toFixed(3)})`,
      });
    }
    // Layer ids must be unique.
    const ids = new Set<string>();
    for (const layer of doc.layers) {
      if (ids.has(layer.id)) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          path: ['layers'],
          message: `duplicate layer id: ${layer.id}`,
        });
      }
      ids.add(layer.id);
    }
  });
export type SceneDocument = z.infer<typeof sceneDocumentSchema>;

/**
 * Parse and validate an unknown value as a SceneDocument. Throws a ZodError with
 * readable messages on failure.
 */
export function parseSceneDocument(input: unknown): SceneDocument {
  return sceneDocumentSchema.parse(input);
}

/** Non-throwing variant returning a discriminated result. */
export function safeParseSceneDocument(input: unknown) {
  return sceneDocumentSchema.safeParse(input);
}

File: apps/ai-service/schemas/scene.py
"""Pydantic v2 mirror of the TypeScript Zod scene schema.

Field names round-trip with the TypeScript side: models use a camelCase alias
generator and accept both snake_case (Python) and camelCase (JSON) input. Dump
with ``model_dump(by_alias=True)`` to produce TS-compatible JSON.

Mirrors ``packages/scene-schema/src/scene.ts`` and ``audio.ts``.
"""

from __future__ import annotations

from typing import Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    model_validator,
)
from pydantic.alias_generators import to_camel

SCENE_SCHEMA_VERSION = "1.0"
ASPECT_RATIO_TOLERANCE = 0.02

PresetName = Literal["soft-nature", "urban", "dark", "nostalgic"]
LayerRole = Literal[
    "background",
    "distant",
    "midground",
    "secondary-subject",
    "primary-subject",
    "foreground",
    "atmosphere",
]
ReturnMode = Literal["spring", "inertia", "physics", "fixed"]
VisiblePixels = Literal["original", "generated", "mixed"]
SegmentationDifficulty = Literal["low", "medium", "high"]
AudioBindingSource = Literal["bass", "lowMid", "highMid", "treble", "loudness", "beatPulse"]
AudioCurve = Literal["linear", "easeIn", "easeOut", "exponential"]


class SceneModel(BaseModel):
    """Base model: camelCase JSON aliases, accepts snake_case or camelCase input."""

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )


class Bounds(SceneModel):
    """Normalized [0,1] rectangle in image space (origin top-left)."""

    x: float
    y: float
    width: float = Field(gt=0)
    height: float = Field(gt=0)


class Anchor(SceneModel):
    x: float
    y: float


class LayerMovement(SceneModel):
    enabled: bool
    max_offset_x: float = Field(ge=0)
    max_offset_y: float = Field(ge=0)
    max_offset_z: float = Field(ge=0)
    max_rotation: float = Field(ge=0)
    parallax_strength: float = Field(ge=0)
    drag_enabled: bool
    return_mode: ReturnMode


class AudioSensitivity(SceneModel):
    bass: float = Field(ge=0, le=1)
    low_mid: float = Field(ge=0, le=1)
    high_mid: float = Field(ge=0, le=1)
    treble: float = Field(ge=0, le=1)
    beat: float = Field(ge=0, le=1)
    loudness: float = Field(ge=0, le=1)


class RevealBudget(SceneModel):
    """Limits on how far a layer may move, so unreconstructed pixels stay hidden."""

    max_offset_x: float = Field(ge=0)
    max_offset_y: float = Field(ge=0)
    confidence: float = Field(ge=0, le=1)
    generated_coverage_mask_url: str | None = None


class Provenance(SceneModel):
    visible_pixels: VisiblePixels
    source_image_hash: str
    segmentation_provider: str | None = None
    matting_provider: str | None = None
    completion_provider: str | None = None
    provider_request_ids: list[str] | None = None


class SceneLayer(SceneModel):
    id: str = Field(min_length=1)
    name: str
    semantic_label: str
    role: LayerRole

    asset_url: str = Field(min_length=1)
    mask_url: str | None = None
    alpha_asset_url: str | None = None

    bounds: Bounds
    anchor: Anchor

    # Relative depth in [0,1]; 0 = nearest camera, 1 = farthest.
    depth: float = Field(ge=0, le=1)
    depth_variance: float = Field(ge=0)
    base_scale: float = Field(gt=0)
    base_rotation: float
    base_opacity: float = Field(ge=0, le=1)

    movement: LayerMovement

    interaction_tags: list[str]
    material_tags: list[str]
    audio_sensitivity: AudioSensitivity

    importance: float = Field(ge=0, le=1)
    locked: bool

    reveal_budget: RevealBudget
    provenance: Provenance


class VisualAnalysis(SceneModel):
    scene_type: str
    main_subject: str
    secondary_subjects: list[str] = Field(default_factory=list)
    foreground_candidates: list[str] = Field(default_factory=list)
    midground_candidates: list[str] = Field(default_factory=list)
    background: str = ""
    water_regions: list[str] = Field(default_factory=list)
    vegetation_regions: list[str] = Field(default_factory=list)
    sky_regions: list[str] = Field(default_factory=list)
    transparent_or_reflective_regions: list[str] = Field(default_factory=list)
    lighting_direction: str = ""
    depth_of_field: str = ""
    camera_style: str = ""
    dominant_colors: list[str] = Field(default_factory=list)
    mood: str = ""
    suggested_preset: PresetName | None = None
    suggested_motion_tags: list[str] = Field(default_factory=list)
    suggested_segmentation_prompts: list[str] = Field(default_factory=list)
    locked_objects: list[str] = Field(default_factory=list)
    expected_segmentation_difficulty: SegmentationDifficulty = "medium"
    inpainting_risk_areas: list[str] = Field(default_factory=list)


class FogConfig(SceneModel):
    enabled: bool = False
    density: float = Field(default=0, ge=0, le=1)
    color: str = "#ffffff"
    layers: int = Field(default=0, ge=0, le=5)


class AtmosphereConfig(SceneModel):
    fog: FogConfig = Field(default_factory=FogConfig)
    ambient_light: float = Field(default=1, ge=0, le=2)
    vignette: float = Field(default=0, ge=0, le=1)
    grain: float = Field(default=0, ge=0, le=1)


class CameraConfig(SceneModel):
    # Field of view in degrees for the perspective camera.
    fov: float = Field(default=45, ge=1, le=179)
    parallax_strength: float = Field(default=0.16, ge=0)
    zoom_response: float = Field(default=0.03, ge=0)
    shake_strength: float = Field(default=0, ge=0)
    drift_strength: float = Field(default=0, ge=0)


class AudioBinding(SceneModel):
    source: AudioBindingSource
    target: str = Field(min_length=1)
    scale: float
    offset: float
    smoothing: float = Field(ge=0, le=1)
    clamp: tuple[float, float]
    curve: AudioCurve = "linear"


class SceneMetadata(SceneModel):
    created_at: str
    pipeline_version: str


class SceneDocument(SceneModel):
    version: Literal["1.0"]
    id: str = Field(min_length=1)
    title: str
    width: int = Field(gt=0)
    height: int = Field(gt=0)
    aspect_ratio: float = Field(gt=0)
    original_image_url: str = Field(min_length=1)
    background_plate_url: str = Field(min_length=1)
    depth_map_url: str | None = None
    preset: PresetName
    visual_analysis: VisualAnalysis
    layers: list[SceneLayer] = Field(min_length=1)
    atmosphere: AtmosphereConfig
    camera: CameraConfig
    audio_bindings: list[AudioBinding] = Field(default_factory=list)
    metadata: SceneMetadata

    @model_validator(mode="after")
    def _check_refinements(self) -> SceneDocument:
        # Aspect ratio must be internally consistent (within rounding tolerance).
        expected = self.width / self.height
        if abs(expected - self.aspect_ratio) > ASPECT_RATIO_TOLERANCE:
            raise ValueError(
                f"aspectRatio {self.aspect_ratio} does not match "
                f"width/height ({expected:.3f})"
            )
        # Layer ids must be unique.
        seen: set[str] = set()
        for layer in self.layers:
            if layer.id in seen:
                raise ValueError(f"duplicate layer id: {layer.id}")
            seen.add(layer.id)
        return self


def parse_scene_document(data: dict) -> SceneDocument:
    """Parse and validate a dict as a :class:`SceneDocument`.

    Raises ``pydantic.ValidationError`` on failure.
    """
    return SceneDocument.model_validate(data)
